core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -114,42 +115,92 @@ class RangePolicy : public Impl::PolicyTraits<Properties...> {
         m_granularity_mask(0) {}
 
   /** \brief  Total range */
+#if !defined(KOKKOS_ENABLE_DEPRECATED_CODE_4) || \
+    defined(KOKKOS_ENABLE_DEPRECATION_WARNINGS)
+  template <typename IndexType1, typename IndexType2,
+            std::enable_if_t<(std::is_convertible_v<IndexType1, member_type> &&
+                              std::is_convertible_v<IndexType2, member_type>),
+                             bool> = false>
+  inline RangePolicy(const typename traits::execution_space& work_space,
+                     const IndexType1 work_begin, const IndexType2 work_end)
+#else
   inline RangePolicy(const typename traits::execution_space& work_space,
                      const member_type work_begin, const member_type work_end)
+#endif
       : m_space(work_space),
         m_begin(work_begin),
         m_end(work_end),
         m_granularity(0),
         m_granularity_mask(0) {
+#ifdef KOKKOS_ENABLE_DEBUG
+    check_conversion_safety(work_begin);
+    check_conversion_safety(work_end);
+#endif
     check_bounds_validity();
     set_auto_chunk_size();
   }
 
   /** \brief  Total range */
+#if !defined(KOKKOS_ENABLE_DEPRECATED_CODE_4) || \
+    defined(KOKKOS_ENABLE_DEPRECATION_WARNINGS)
+  template <typename IndexType1, typename IndexType2,
+            std::enable_if_t<(std::is_convertible_v<IndexType1, member_type> &&
+                              std::is_convertible_v<IndexType2, member_type>),
+                             bool> = false>
+  inline RangePolicy(const IndexType1 work_begin, const IndexType2 work_end)
+#else
   inline RangePolicy(const member_type work_begin, const member_type work_end)
-      : RangePolicy(typename traits::execution_space(), work_begin, work_end) {}
+#endif
+      : RangePolicy(typename traits::execution_space(), work_begin, work_end) {
+  }
 
   /** \brief  Total range */
+#if !defined(KOKKOS_ENABLE_DEPRECATED_CODE_4) || \
+    defined(KOKKOS_ENABLE_DEPRECATION_WARNINGS)
+  template <typename IndexType1, typename IndexType2, typename... Args,
+            std::enable_if_t<(std::is_convertible_v<IndexType1, member_type> &&
+                              std::is_convertible_v<IndexType2, member_type>),
+                             bool> = false>
+  inline RangePolicy(const typename traits::execution_space& work_space,
+                     const IndexType1 work_begin, const IndexType2 work_end,
+                     Args... args)
+#else
   template <class... Args>
   inline RangePolicy(const typename traits::execution_space& work_space,
                      const member_type work_begin, const member_type work_end,
                      Args... args)
+#endif
       : m_space(work_space),
         m_begin(work_begin),
         m_end(work_end),
         m_granularity(0),
         m_granularity_mask(0) {
+#ifdef KOKKOS_ENABLE_DEBUG
```

Comment:
Why did you guard for `KOKKOS_ENABLE_DEBUG `?
I think we want to enable only based on `KOKKOS_ENABLE_DEPRECATED_CODE_4` and `KOKKOS_ENABLE_DEPRECATION_WARNINGS`.
`KOKKOS_ENABLE_DEBUG` is an "opt-in" option but we would really want users to get the runtime warnings unless they disable warnings.

Reply:
Removed this guard and in the unit test.

> Also the function for the representability check should only be called in debug mode maybe?

It was added in attempt to address this feedback.